internal/rpcserver/rpcserver.go

```diff
@@ -6070,7 +6097,20 @@ func (s *Server) route(ctx context.Context) *http.Server {
 		// default size for read/write buffers and impose a read limit that
 		// depends on whether or not the connection is authenticated yet.
 		upgrader := websocket.Upgrader{
-			CheckOrigin: func(r *http.Request) bool { return true },
+			CheckOrigin: func(r *http.Request) bool {
+				origin := r.Header["Origin"]
+				if len(origin) == 0 {
+					return true
+				}
+				u, err := url.Parse(origin[0])
+				if err != nil {
+					return false
+				}
+				if len(u.Host) == 0 {
+					return true
+				}
+				return equalASCIIFold(u.Host, r.Host)
```

Comment:
Only thing I'm wondering about this is if we want to strip the port out of `Host` and only compare the actual host. Otherwise I think the strict check is including port, which would never be equal even if you had some html served from a localhost IP.  https://github.com/gorilla/websocket/issues/592

Reply:
The `if len(u.Host) == 0 {` part is meant to handle a `file:///` resource?

Reply:
e.g.

```go
// Origin is local webserver
u, err := url.Parse("http://localhost:1234")
if err != nil {
fmt.Println(err)
return
}
oH, oP, _ := net.SplitHostPort(u.Host)
fmt.Println(oH, oP) // localhost 1234

// Request's Host is dcrd's RPC server
reqHost := "localhost:9109"
rH, rP, _ := net.SplitHostPort(reqHost)
fmt.Println(rH, rP) // localhost 9109

fmt.Println("u.Host == r.Host", u.Host == reqHost) // false
fmt.Println("oH == rH", oH==rH) // true
```

Reply:
> The `if len(u.Host) == 0 {` part is meant to handle a `file:///` resource where some browser still set an Origin, to something like `null`?

Correct.  However, I went ahead and modified the code a bit to be more explicit in looking exactly for the `null` path set by Firefox and the `file` scheme set by Chrome, and Edge and added some comments to make it more explicit and clear what it's doing.



> Only thing I'm wondering about this is if we want to strip the port out of `Host` and only compare the actual host. Otherwise I think the strict check is including port, which would never be equal even if you had some html served from a localhost IP.

Good point.  I've updated the code to strip the port from the origin and request hosts accordingly.  I've also added some comments throughout to better explain the intent.